RN-173 Added support for Markdown images

hmhealey opened the pull request:
This one is a bit more complicated than the regular markdown changes because it requires moving images out of paragraphs, lists, etc because React Native doesn't support nesting them within text except under very specific circumstances (that emoji have to follow). This means that if you had something like
```
This is a sentence with ![an image](http://example.com/image) in it
```
it's instead converted into something more like
```
This is a sentence with
![an image](http://example.com/image)
in it
```

I added Asaad to this in case there's any changes that should be made to the appearance of the images or the text that's displayed when they cannot be loaded. Also, if you want to test this yourself, here's some test images you can use:
```
![square](https://github.com/hmhealey/test-files/blob/master/red-10000x10000.png?raw=true)
![tall](https://github.com/hmhealey/test-files/blob/master/green-2000x10000.png?raw=true)
![wide](https://github.com/hmhealey/test-files/blob/master/blue-10000x2000.png?raw=true)
```

#### Ticket Link
https://mattermost.atlassian.net/browse/RN-173

#### Checklist
- Added or updated unit tests (required for all new features)
- Has UI changes
- Includes text changes and localization file updates

#### Device Information
This PR was tested on: iOS Simulator (iPhone 6, iOS 11.1), Nexus 5 (Android 6.0.1)

#### Screenshots
<img width="390" alt="screen shot 2017-11-24 at 11 20 10 am" src="https://user-images.githubusercontent.com/3277310/33218832-fab46f06-d10c-11e7-95f4-b4e61f1cd223.png">
<img width="381" alt="screen shot 2017-11-24 at 11 51 39 am" src="https://user-images.githubusercontent.com/3277310/33219146-c8fedc60-d10e-11e7-9b43-31ef8b900b04.png">
<img width="383" alt="screen shot 2017-11-24 at 11 19 44 am" src="https://user-images.githubusercontent.com/3277310/33218834-006ab932-d10d-11e7-91da-0939b2408452.png">
<img width="379" alt="screen shot 2017-11-24 at 11 59 05 am" src="https://user-images.githubusercontent.com/3277310/33219177-ec7314c2-d10e-11e7-862e-f424ae73154d.png">

hmhealey commented:
I've added a couple changes to
1. Fix the alignment of text that comes after an image. Previously, "This is ![an image](...) in a sentence" would result in the " in a sentence" text keeping the space at the start of it. Since it was on its own line, this caused it to appear to be indented slightly, so that space is now removed.
2. Images now work in links. Previously, the link URL would pave over the image URL because I mixed up some names, and it had to click handler for the link to actually work
3. I fixed some changes that caused a previous bug to re-appear where upper case links on Android didn't work

enahum commented on app/components/markdown/markdown_image.js:
```diff
@@ -1,70 +1,206 @@
+    handleLinkPress = () => {
+        const url = normalizeProtocol(this.props.linkDestination);
+
+        Linking.canOpenURL(url).then((supported) => {
```
is this expected? do we do this on the webapp?

hmhealey replied:
This only happens if their `linkDestination` prop is set which only occurs when the image is inside of a link. This is here because we can't do what the web app does when it turns `[![Logo](https://example.com/logo)](https://example.com)` into `<a href="https://example.com"><img src="https://exampl.ecom/logo"/></a>`

enahum commented on app/components/markdown/markdown_image.js:
```diff
@@ -1,70 +1,206 @@
+        let action;
+        if (config.copyAndPasteProtection !== 'true') {
+            action = {
+                text: formatMessage({id: 'mobile.markdown.link.copy_url', defaultMessage: 'Copy URL'}),
```
this suppose to be an image right? why not set the menu item as `Copy Image URL` ?

hmhealey replied:
This is for the same behaviour as above since this image is part of a link

hmhealey commented:
@asaadmahmood I added a 5 pixel vertical margin around the images and fixed another issue with spacing around images, so here's what it looks like currently with and without surrounding text
<img width="399" alt="screen shot 2017-11-27 at 5 59 26 pm" src="https://user-images.githubusercontent.com/3277310/33296624-965d0bee-d3aa-11e7-8a20-cc4a0ea179af.png">

For the error text, it is actually on a new line, so other than that, I added some colour to the text to make it stand out more.
![image](https://user-images.githubusercontent.com/3277310/33296737-12d6ec30-d3ab-11e7-932b-3148b9a858cd.png)

Also, for everyone else, I did some more testing and found out that my Android phone can't render images over a certain size (and crashes if you even try to load images past a larger size), so I added the above error message. I have it set to 4096 by 4096 right now since that matches my phone, but I'm going to look into querying that size. Hopefully [this link](https://stackoverflow.com/questions/7428996/hw-accelerated-activity-how-to-get-opengl-texture-size-limit) works, otherwise I need to query OpenGL directly

enahum commented on app/components/markdown/markdown_image.js:
```diff
@@ -1,70 +1,231 @@
+        if (this.state.width && this.state.height && this.state.maxWidth) {
+            let {width, height} = this.state;
+
+            if (Platform.OS === 'android' && (width > ANDROID_MAX_WIDTH || height > ANDROID_MAX_HEIGHT)) {
```
I'm ok with this for now, but maybe we should look at resizing the image somehow

asaadmahmood commented:
Cool, looks good to me.
